cpp/include/cuvs/neighbors/batch_ann.hpp

```diff
@@ -0,0 +1,194 @@
+   *
+   * @param res raft::resources is an object mangaging resources
+   * @param n_rows number of rows in knn-graph
+   * @param k number of nearest neighbors in knn-graph
+   * @param return_distances whether to return distances
+   */
+  index(raft::resources const& res, int64_t n_rows, int64_t k, bool return_distances = false)
+    : cuvs::neighbors::index(),
+      res_{res},
+      k_{k},
+      graph_{raft::make_host_matrix<IdxT, IdxT, raft::row_major>(n_rows, k)},
+      graph_view_{graph_.view()},
+      return_distances_{return_distances}
+  {
+    if (return_distances) {
+      distances_      = raft::make_device_matrix<DistT, IdxT>(res, n_rows, k);
+      distances_view_ = distances_.value().view();
+    }
+  }
+
+  /**
+   * @brief Construct a new index object
+   *
+   * This constructor creates an all-neighbors graph using a user allocated host memory knn-graph.
+   * The type of the knn-graph is a dense raft::host_matrix and dimensions are (n_rows, k).
+   *
+   * @param res raft::resources is an object mangaging resources
+   * @param graph_view raft::host_matrix_view<IdxT, IdxT, raft::row_major> for storing knn-graph
+   * @param distances_view optional raft::device_matrix_view<DistT, IdxT, row_major> for storing
+   * distances
+   */
+  index(
+    raft::resources const& res,
+    raft::host_matrix_view<IdxT, IdxT, raft::row_major> graph_view,
+    std::optional<raft::device_matrix_view<DistT, IdxT, row_major>> distances_view = std::nullopt)
+    : cuvs::neighbors::index(),
+      res_{res},
+      k_{graph_view.extent(1)},
+      graph_{raft::make_host_matrix<IdxT, IdxT, raft::row_major>(0, 0)},
+      graph_view_{graph_view},
+      distances_view_{distances_view},
+      return_distances_{distances_view.has_value()}
+  {
+  }
+
+  bool return_distances() { return return_distances_; }
+
+  int64_t k() { return k_; }
+
+  /** neighborhood graph [size, k] */
+  [[nodiscard]] inline auto graph() noexcept -> raft::host_matrix_view<IdxT, IdxT, raft::row_major>
+  {
+    return graph_view_;
+  }
+
+  /** neighborhood graph distances [size, k] */
+  [[nodiscard]] inline auto distances() noexcept
+    -> std::optional<device_matrix_view<DistT, IdxT, row_major>>
+  {
+    return distances_view_;
+  }
+
+  // Don't allow copying the index for performance reasons (try avoiding copying data)
+  index(const index&)                    = delete;
+  index(index&&)                         = default;
+  auto operator=(const index&) -> index& = delete;
+  auto operator=(index&&) -> index&      = default;
+  ~index()                               = default;
+
+ private:
+  raft::resources const& res_;
+  int64_t k_;
+  bool return_distances_;
+  raft::host_matrix<IdxT, IdxT, raft::row_major> graph_;
+  std::optional<raft::device_matrix<DistT, IdxT, row_major>> distances_;
+  raft::host_matrix_view<IdxT, IdxT, raft::row_major> graph_view_;
+  std::optional<raft::device_matrix_view<DistT, IdxT, row_major>> distances_view_;
```

Comment:
Same here, if we store the owning structure, we can create a view on the fly which is super cheap (it's literally just creating an object wrapped around the array in memory.

Reply:
(comment above)